data/transactions/logic/eval_test.go

```diff
@@ -1806,6 +1807,178 @@ func testLogic(t *testing.T, program string, v uint64, ep EvalParams, problems .
 	}
 }
 
+func TestGtxnaScratch(t *testing.T) {
+	t.Parallel()
+
+	type scratchTestCase struct {
+		tealSources []string
+		errContains string
+	}
+
+	simpleCase := scratchTestCase{
+		tealSources: []string{
+			`int 2
+store 0
+int 1`,
+			`gtxna 0 Scratch 0
+int 2
+==
+`,
+		},
+	}
+
+	multipleTxnCase := scratchTestCase{
+		tealSources: []string{
+			`byte "txn 1"
+store 0
+int 1`,
+			`byte "txn 2"
+store 1
+int 1`,
+			`gtxna 0 Scratch 0
+byte "txn 1"
+==
+gtxna 1 Scratch 1
+byte "txn 2"
+==
+&&
+`,
+		},
+	}
+
+	selfCase := scratchTestCase{
+		tealSources: []string{
+			`gtxna 0 Scratch 0
+int 2
+store 0
+int 1
+`,
+		},
+		errContains: "can't use Scratch txn field on self, use load instead",
+	}
+
+	laterTxnSlotCase := scratchTestCase{
+		tealSources: []string{
+			`gtxna 1 Scratch 0
+int 2
+==
+`,
+			`int 2
+store 0
+int 1`,
+		},
+		errContains: "can't get future Scratch from txn with index 1",
+	}
+
+	cases := []scratchTestCase{
+		simpleCase, multipleTxnCase, selfCase, laterTxnSlotCase,
+	}
+	proto := defaultEvalProtoWithVersion(LogicVersion)
+
+	for i, testCase := range cases {
+		t.Run(fmt.Sprintf("i=%d", i), func(t *testing.T) {
+			sources := testCase.tealSources
+			// Assemble ops
+			opsList := make([]*OpStream, len(sources))
+			for j, source := range sources {
+				ops, err := AssembleStringWithVersion(source, AssemblerMaxVersion)
```

Comment:
Use testProg when possible.  It shows what's going on if your assembly fails, and eliminates the need to test the error.